frontend/src/utils/PrivateRoute.tsx

```diff
@@ -1,52 +1,68 @@
-import { Redirect, Route, RouteProps, useLocation } from 'react-router-dom';
+import { Redirect, Route, RouteComponentProps, RouteProps, useLocation } from 'react-router-dom';
 
-interface IPrivateRouteProps extends RouteProps {
-  component: React.ComponentType<any>;
+interface BaseProtectedAppRoute extends RouteProps {
+  role?: string | string[];
+  claim?: string | string[];
+}
+interface ComponentRoute extends BaseProtectedAppRoute {
+  customComponent: React.ComponentType<any>;
   layout: React.ComponentType<any>;
-  role?: string | Array<string>;
-  claim?: string | Array<string>;
-  componentProps?: any;
+}
+
+interface RenderRoute extends BaseProtectedAppRoute {
+  customRender: (props: RouteComponentProps<any>) => React.ReactNode;
+}
+
+type IPrivateRouteProps = ComponentRoute | RenderRoute;
+
+function isRenderRoute(route: RenderRoute | ComponentRoute): route is RenderRoute {
+  return (route as RenderRoute).customRender !== undefined;
 }
 
 /**
  * A PrivateRoute only allows a user who is authenticated and has the appropriate role(s) or claim(s).
  * @param props - Properties to pass { component, role, claim }
  */
-const PrivateRoute = (props: IPrivateRouteProps) => {
+const PrivateRoute: React.FC<IPrivateRouteProps> = props => {
   const location = useLocation();
   const keycloak = useKeycloakWrapper();
-  let { component: Component, layout: Layout, ...rest } = props;
-  return (
-    <Route
-      {...rest}
-      render={routeProps => {
-        if (!!keycloak.obj?.authenticated) {
-          if (
-            (!rest.role && !rest.claim) ||
-            keycloak.hasRole(rest.role) ||
-            keycloak.hasClaim(rest.claim)
-          ) {
-            return (
+  let { ...rest } = props;
+
+  if (!!keycloak.obj?.authenticated) {
+    if (
+      (!rest.role && !rest.claim) ||
+      keycloak.hasRole(rest.role) ||
+      keycloak.hasClaim(rest.claim)
+    ) {
+      if (isRenderRoute(props)) {
+        return <Route {...rest} render={props.customRender} />;
+      } else {
+        const { customComponent: Component, layout: Layout } = props;
+        return (
+          <Route
+            {...rest}
+            render={props => (
               <Layout>
-                <Component {...routeProps} {...rest.componentProps} />
+                <Component {...props} />
               </Layout>
-            );
-          } else {
-            return (
-              <Redirect to={{ pathname: '/forbidden', state: { referer: routeProps.location } }} />
-            );
-          }
-        } else {
-          if (routeProps.location.pathname !== '/login') {
-            const redirectTo = encodeURI(`${location.pathname}${location.search}`);
-            return <Redirect to={`/login?redirect=${redirectTo}`} />;
-          }
-        }
-      }}
-    />
-  );
+            )}
+          />
+        );
+      }
+    } else {
+      return <Redirect to={{ pathname: '/forbidden', state: { referer: location } }} />;
```

Comment:
hmm, I'm worried that this may not always be equivalent. routeProps.location = the location object from the component. location = the most up to date location object.

Reply:
Uhm, I dont know either. will maintain the same for now